File: Front/src/app/services/notificacion.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

/** Mismos valores que el enum TipoNotificacion del backend. */
export type TipoNotificacion = 'CAJA_ATRASADA' | 'RECHAZO_OPERACION';

/** Qué tipos tienen algo pendiente de avisarle al usuario logueado ahora mismo. */
export type ResumenNotificaciones = Partial<Record<TipoNotificacion, boolean>>;

@Injectable({
  providedIn: 'root',
})
export class NotificacionService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiBase}/interno/notificaciones`;

  /** Qué tipos de aviso tienen pendientes ahora mismo, para el usuario logueado. ADMIN-only. */
  obtenerResumen(): Observable<ResumenNotificaciones> {
    return this.http.get<ResumenNotificaciones>(`${this.apiUrl}/resumen`);
  }

  /** Sólo tiene efecto real en tipos con desapareceAlVerse=true (ej. CAJA_ATRASADA): en los
   * demás (ej. RECHAZO_OPERACION) el aviso sólo se apaga resolviendo la entidad referida.
   * Devuelve los ids que estaban SIN ver hasta ahora (los que prendían el aviso), para señalarlos
   * en pantalla: una vez marcados, el aviso se apaga y ya no hay otra forma de saber cuáles eran. */
  marcarVistas(tipo: TipoNotificacion, refIds: number[]): Observable<number[]> {
    return this.http.post<number[]>(`${this.apiUrl}/${tipo}/marcar-vistas`, { refIds });
  }
}

File: Back/La Serranita entradas/src/main/java/org/example/laserranitaentradas/service/NotificacionService.java
package org.example.laserranitaentradas.service;

import org.example.laserranitaentradas.model.entity.TipoNotificacion;

import java.util.List;

/**
 * Sistema genérico de avisos (punto rojo en la cabecera interna): cada feature (cajas atrasadas,
 * rechazos pendientes, lo que se agregue a futuro) calcula su propia lista de "candidatos"
 * vigentes (ids de la entidad que sea) y este servicio decide si eso cuenta como pendiente para
 * un usuario puntual, según si el tipo se apaga con sólo verlo (TipoNotificacion.
 * desapareceAlVerse) o exige una acción explícita sobre la entidad referida.
 */
public interface NotificacionService {

    /** true si, de los candidatos vigentes, hay algo pendiente de avisarle a este usuario. */
    boolean hayPendientes(TipoNotificacion tipo, List<Long> refIdsActuales, Long usuarioId);

    /** Marca estos ids como vistos por este usuario para este tipo. Sólo tiene efecto real en
     * tipos con desapareceAlVerse=true; en los demás no rompe nada, simplemente no se consulta.
     *
     * Devuelve los ids que hasta este momento estaban SIN ver para este usuario, o sea los que
     * estaban prendiendo el aviso: la pantalla los usa para señalar cuáles son las novedades entre
     * todos los candidatos (ej. cuál de las cajas atrasadas es la nueva). Vacío si no había nada
     * nuevo o el tipo no se apaga al verse. */
    List<Long> marcarVistas(TipoNotificacion tipo, List<Long> refIds, Long usuarioId);
}
